core/src/integration/java/com/radixdlt/integration/targeted/consensus/EpochTimeoutCertTest.java

```diff
@@ -0,0 +1,132 @@
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.integration.targeted.consensus;
+
+import static com.radixdlt.harness.deterministic.invariants.DeterministicMonitors.byzantineBehaviorNotDetected;
+
+import com.radixdlt.consensus.EpochNodeWeightMapping;
+import com.radixdlt.consensus.Proposal;
+import com.radixdlt.consensus.bft.Round;
+import com.radixdlt.environment.deterministic.network.ControlledMessage;
+import com.radixdlt.environment.deterministic.network.MessageSelector;
+import com.radixdlt.harness.deterministic.DeterministicTest;
+import com.radixdlt.harness.deterministic.PhysicalNodeConfig;
+import com.radixdlt.modules.FunctionalRadixNodeModule;
+import com.radixdlt.modules.StateComputerConfig;
+import io.reactivex.rxjava3.schedulers.Timed;
+import java.util.function.Predicate;
+import org.junit.Test;
+
+public final class EpochTimeoutCertTest {
+
+  private static final long ROUNDS_PER_EPOCH = 100;
+
+  private DeterministicTest createTest() {
+    return DeterministicTest.builder()
+        .addPhysicalNodes(PhysicalNodeConfig.createBasicBatch(4))
+        .messageSelector(MessageSelector.firstSelector())
+        .messageMutator(
+            (m, q) -> {
+              // Drop single proposal at end of epoch to cause a TC
```

Comment:
Just to check - was this test broken before your fix?

In its current formulation, I think it causes everyone to get a TC for round 102, so everyone would commit the same ledger header in round (say, 105?)

Can we have tests for both scenarios:
* A timeout in round 101/102 (everyone should progress to epoch at round 105, say)
* Proposal with QC to commit epoch change dropped in round 103 to people other than the leader and maybe 1 other. The leader would go to the next epoch before the others.

Actually, I think we can still get a liveness break if we're unlucky - if we keep getting a repeat of a leader going to the next epoch in (say) rounds 103 and rounds 106 - if the leaders drop out of consensus for epoch N when they progress to epoch N + 1? So can we have a third test for that too?

Reply:
Actually maybe this isn't an issue because round 103 would timeout, people would resend their votes, and everyone would form a timeout vertex QC...

Reply:
BUT if the proposal _did_ go to a few people, nodes would form a TC first, and advance to the next round without forming the timeout vertex QC

Reply:
If half the validator set form the TC and advance without committing hte epoch change; and hte other half are happily on the next epoch, we have a problem.